lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -13,45 +13,369 @@
+    for (auto portInfo : component.getPortInfo())
+      hwInputInfo.push_back(
+          {portInfo.name, hwDirection(portInfo.direction), portInfo.type});
+
+    auto hwMod = rewriter.create<HWModuleOp>(
+        component.getLoc(), component.getNameAttr(), hwInputInfo);
+
+    rewriter.eraseOp(hwMod.getBodyBlock()->getTerminator());
+    ConversionPatternRewriter::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointToEnd(hwMod.getBodyBlock());
+
+    SmallVector<Value> argValues;
+    SmallVector<Value> outputWires;
+    size_t portIdx = 0;
+    for (auto portInfo : component.getPortInfo()) {
+      switch (portInfo.direction) {
+      case calyx::Direction::Input:
+        assert(hwMod.getArgument(portIdx).getType() == portInfo.type);
+        argValues.push_back(hwMod.getArgument(portIdx));
+        break;
+      case calyx::Direction::Output:
+        auto wire = rewriter.create<sv::WireOp>(component.getLoc(),
+                                                portInfo.type, portInfo.name);
+        auto wireRead =
+            rewriter.create<sv::ReadInOutOp>(component.getLoc(), wire);
+        argValues.push_back(wireRead);
+        outputWires.push_back(wireRead);
+        break;
+      }
+      ++portIdx;
     }
-  };
 
-  addHWPortInfo(op.getInputPortInfo(), hw::PortDirection::INPUT);
-  addHWPortInfo(op.getOutputPortInfo(), hw::PortDirection::OUTPUT);
+    rewriter.mergeBlocks(component.getBody(), hwMod.getBodyBlock(), argValues);
+    rewriter.create<OutputOp>(component.getLoc(), outputWires);
+    rewriter.eraseOp(component);
 
-  return builder.create<hw::HWModuleExternOp>(
-      op->getLoc(), builder.getStringAttr(op.getName()), hwPortInfos);
-}
+    return success();
+  }
+
+private:
+  hw::PortDirection hwDirection(calyx::Direction dir) const {
+    switch (dir) {
+    case calyx::Direction::Input:
+      return hw::PortDirection::INPUT;
+    case calyx::Direction::Output:
+      return hw::PortDirection::OUTPUT;
+    }
+  }
+};
+
+struct ConvertWiresOp : public OpConversionPattern<WiresOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(WiresOp wires, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    HWModuleOp hwMod = wires->getParentOfType<HWModuleOp>();
+    rewriter.inlineRegionBefore(wires.body(), hwMod.getBodyRegion(),
+                                hwMod.getBodyRegion().end());
+    rewriter.eraseOp(wires);
+    rewriter.mergeBlockBefore(&hwMod.getBodyRegion().getBlocks().back(),
+                              &hwMod.getBodyBlock()->back());
+    return success();
+  }
+};
+
+struct ConvertControlOp : public OpConversionPattern<ControlOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ControlOp control, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(control.getBody()->empty() && "calyx control must be structural");
```

Comment:
Should this be an assert? I can see people calling this pass with potentially unlowered Calyx code, and in that case, I'd rather have an error message than the compiler crashing.

Reply:
+1

Reply:
I'd argue that assertions are for telling compiler engineers that an invariant of the compiler is broken, and errors are for telling end-users the compiler couldn't handle their input. IMHO this is firmly in compiler invariant territory: "people calling this pass" aren't end-users, they are compiler engineers. Users use something like hlstool, not circt-opt.

All that being said, MLIR guidelines dictate no pass should assert on verified IR, and the Calyx IR allows wires to be empty or not at any point in any pipeline, so I can make this an error.

Reply:
The way I see it, our goal is to make the compiler user-friendly by reducing friction where possible. The legal state of Calyx IR before lowering is probably not well understood by many compiler engineers today. This is different than, say, "you passed in a zero-element container and tried to access index 0." 

Having an error message with what steps need to be taken to avoid this, i.e., exporting to the Rust compiler to compile control, is much nicer than "calyx control must be structural," which isn't informative unless you're familiar with Calyx.